src/udev/net/ethtool-util.c

```diff
@@ -206,3 +212,116 @@ int ethtool_set_wol(int *fd, const char *ifname, WakeOnLan wol) {
 
         return 0;
 }
+
+static int ethtool_get_stringset(int *fd, struct ifreq *ifr, int stringset_id, struct ethtool_gstrings **gstrings) {
+        _cleanup_free_ struct ethtool_sset_info *info = NULL;
+        _cleanup_free_ struct ethtool_gstrings *strings = NULL;
+        unsigned len;
+        int r;
+
+        info = malloc0(sizeof (*info) + sizeof (unsigned));
+        if (!info)
+                return log_oom();
+
+        info->cmd = ETHTOOL_GSSET_INFO;
+        info->reserved = 0;
+        info->sset_mask = 1ULL << stringset_id;
+
+        ifr->ifr_data = (void *) info;
+
+        r = ioctl(*fd, SIOCETHTOOL, ifr);
+        if (r < 0)
+                return -errno;
+
+        if (!info->sset_mask)
+                return -1;
+
+        len = info->data[0];
+
+        strings = malloc0(sizeof(*strings) + len * ETH_GSTRING_LEN);
+        if (!strings)
+                return log_oom();
+
```

Comment:
hmm, so the above would probably be prettier if you'd use alloca0() instead of malloc0(), after all the size is bounded and comes from a trusted source, and we never need to pass this memory outside of the function...